### PR TITLE
November 2018 activation code

### PR DESCRIPTION
**Pull request by sickpig**
Summary:

- Activation code for May 2018 has been removed completely since all the changes (blk size, new op_codes) where never used before the fork. 

- EB default value changed to 32MB, before it was changed during init step

- EB value can't be lower than 32MB, this was the case also before these set of changes. But the constraint was enforced during `UpdateTip` execution. With code in this PR you can't even set the value of this tweak lower then 32MB, cause the validator will raise an error. 

- MG (max generated block size) default value changed to 8MB. I'm open to set it back to 2MB, I chose to change it because in the may 2018 activation code the value of MG was changed to 8MB. Actually there was an a little bit of ambiguity cause client that were on during the upgrade got the value set to 8MB, if after the fork it got rebooted for whatever reason MG got it back to 2MB.

TODO:

- ~~need to fix `validateblocktemplate.py` test, I temporarily disabled a little chunk of the test due to the fact that it is not possible to set EB lower than 32MB anymore.~ (done) 

- ~~excessive.py need to be fixed~~ (done) 

- ~~need to find out block height for testnet may 2018 fork.~~

### CODE REVIEW

**Comment by sickpig** [comment]
all tests fixed, taking out of WIP

File: qa/rpc-tests/excessive.py [hunk 1]
```diff
@@ -74,15 +74,6 @@ def setup_network(self, split=False):
                 pdb.set_trace()
 
     def run_test(self):
-        # Temporary workaround for the may15th hardfork. These tests will not work if the fork is enabled
-        # because we are forcing an EB of 32MB after each block is mined which invalidates many of these tests.
-        # So we set the mocktime to sometime far in advance of the fork.
-        MAY152018_START_TIME = 1526400000;
-        self.nodes[0].setmocktime(MAY152018_START_TIME - 10000)
-        self.nodes[1].setmocktime(MAY152018_START_TIME - 10000)
```
Comment (awemany) [inline qa/rpc-tests/excessive.py:82]: A `for` loop saves two lines of code here...

File: qa/rpc-tests/excessive.py [hunk 2]
```diff
@@ -210,14 +201,14 @@ def testExcessiveSigops(self):
         logging.info("testExcessiveSigops: Cleaning up node state")
 
         # We are not testing excessively sized blocks so make these large
-        self.nodes[0].set("net.excessiveBlock=5000000")
-        self.nodes[1].set("net.excessiveBlock=5000000")
-        self.nodes[2].set("net.excessiveBlock=5000000")
-        self.nodes[3].set("net.excessiveBlock=5000000")
         self.nodes[0].setminingmaxblock(5000000)
         self.nodes[1].setminingmaxblock(5000000)
         self.nodes[2].setminingmaxblock(5000000)
```
Comment (awemany) [inline qa/rpc-tests/excessive.py]: Same comment about `for` loop here...

File: src/chainparams.cpp [hunk 3]
```diff
@@ -133,7 +133,9 @@ class CMainParams : public CChainParams
         // Nov, 13 2017 hard fork
         consensus.daaHeight = 504031;
         // May, 15 2018 hard fork
-        consensus.may2018activationTime = 1526400000;
+        consensus.may2018Height = 530359;
+        // May, 15 2018 hard fork
+        consensus.nov2018ActivationTime = 1542300000;
```
Comment (awemany) [inline src/chainparams.cpp:138]: This can be similarly factored out, please have a look at #1311.

File: src/main.cpp [hunk 4]
```diff
@@ -839,10 +837,6 @@ static bool AcceptToMemoryPoolWorker(CTxMemPool &pool,
 
         // Set extraFlags as a set of flags that needs to be activated.
         uint32_t extraFlags = 0;
-        if (hasMay152018)
```
Comment (awemany) [inline src/main.cpp]: @gandrewstone told me to use a `CTweak` just for CDS/V - but wouldn't it make the most sense to add a SCRIPT_ENABLE_NOV2018_OPCODES flag and use the code here in the same way as for the may upgrade?

**Review by awemany — CHANGES_REQUESTED** [review]

File: qa/rpc-tests/excessive.py (hunk 1 above)
Comment (sickpig) [reply qa/rpc-tests/excessive.py:82]: how so? we removed this 6 lines.

File: qa/rpc-tests/excessive.py (hunk 2 above)
Comment (sickpig) [reply qa/rpc-tests/excessive.py]: ok, makes sense. will do

File: src/main.cpp (hunk 4 above)
Comment (sickpig) [reply src/main.cpp]: I think @gandrewstone told you to that while waiting for this PR to be finalized. I think that you can use the `IsNov152018Enabled()` helper and do it the same way we did in May.

File: src/chainparams.cpp (hunk 3 above)
Comment (sickpig) [reply src/chainparams.cpp:138]: yes we could do it, wonder if it is worth doing cause we are going to change it again just after Nov activation to transform it into an height based param. Once we do this we basically have different values for each net...

File: src/main.cpp [hunk 5]
```diff
@@ -839,10 +837,6 @@ static bool AcceptToMemoryPoolWorker(CTxMemPool &pool,
 
         // Set extraFlags as a set of flags that needs to be activated.
         uint32_t extraFlags = 0;
```
Comment (ptschip) [inline src/main.cpp]: do we still need extraflags?

Comment (sickpig) [reply src/main.cpp]: As the code stands we don't. It is a left over from the op_code activation from May 2018 upgrade. Will remove. But I guess @awemany could use it again from op_checkdatasigverify.

**Comment by sickpig** [comment]
Test a full IBD from scratch with these options:

- checkpoints=0  (disable checkpoints)
- checklevel=4   (for each block `ConnectBlock` is performed)

The node with this patch applied sync to current tip correctly. It took 8 hrs 20 min.

File: qa/rpc-tests/excessive.py (hunk 1 above)
Comment (awemany) [reply qa/rpc-tests/excessive.py:82]: Ooops. Yeah, I guess I am colorblind :D

File: src/main.cpp (hunk 4 above)
Comment (awemany) [reply src/main.cpp]: Makes sense, thanks.

**Review by ptschip — APPROVED** [review]

File: src/chainparams.cpp (hunk 3 above)
Comment (awemany) [reply src/chainparams.cpp:138]: I see. I was looking at the may stuff, which sat there just like that, though. No strong opinion either way.

Comment (sickpig) [reply src/chainparams.cpp:138]: me neither. I'm ok either way. Let's see what others think about it.

**Comment by ptschip** [comment]
needs rebase

**Comment by sickpig** [comment]
PR rebased after #1340 went in. Automatic merge wasn't able to solve a conflict in `main.cpp`.

Last time to "solve" is to find out the block height of last testnet fork. It is not necessary to make the code works but I think it is a good thing to have it.

**Comment by sickpig** [comment]
Probably travis would file due to #1348.

**Comment by sickpig** [comment]
Added testnet block height for May' 18 fork.

File: src/globals.cpp [hunk 6]
```diff
@@ -266,20 +266,21 @@ CTweakRef<unsigned int> maxDataCarrierTweak("mining.dataCarrierSize",
...
 CTweak<bool> unsafeGetBlockTemplate("mining.unsafeGetBlockTemplate",
     "Allow getblocktemplate to succeed even if the chain tip is old or this node is not connected to other nodes",
     false);
-
 CTweak<uint64_t> miningForkEB("mining.forkExcessiveBlock",
     "Set the excessive block to this value at the time of the fork.",
-    32000000); // May2018 HF proposed max block size
+    // May2018 HF proposed max block size, no need to change it for the November '18 since we are not going to have an
+    // blopck size increase
+    MIN_EXCESSIVE_BLOCK_SIZE);
```
Comment (gandrewstone) [inline src/globals.cpp]: Since the block size is not changing, I do not think we need miningForkEB and miningForkMG

File: src/chainparams.cpp [hunk 7]
```diff
@@ -427,8 +434,10 @@ class CRegTestParams : public CChainParams
         consensus.uahfHeight = 0;
         // Nov, 13 hard fork is always on on regtest.
         consensus.daaHeight = 0;
+        // May, 15 2018 hard fork is always 0 on regtest
+        consensus.may2018Height = 0;
         // May, 15 2018 hard fork
```
Comment (gandrewstone) [inline src/chainparams.cpp]: comment change to Nov

File: src/chainparams.cpp [hunk 8]
```diff
@@ -133,7 +133,9 @@ class CMainParams : public CChainParams
         // Nov, 13 2017 hard fork
         consensus.daaHeight = 504031;
         // May, 15 2018 hard fork
-        consensus.may2018activationTime = 1526400000;
+        consensus.may2018Height = 530359;
+        // May, 15 2018 hard fork
```
Comment (gandrewstone) [inline src/chainparams.cpp]: change comment to Nov

File: src/chainparams.cpp [hunk 9]
```diff
@@ -259,7 +261,9 @@ class CUnlParams : public CChainParams
         // Nov, 13 hard fork
         consensus.daaHeight = consensus.DifficultyAdjustmentInterval();
         // May, 15 2018 hard fork
-        consensus.may2018activationTime = 1526400000;
+        consensus.may2018Height = 0;
+        // May, 15 2018 hard fork
```
Comment (gandrewstone) [inline src/chainparams.cpp]: change comment to Nov

File: src/test/excessiveblock_test.cpp [hunk 10]
```diff
@@ -20,29 +22,52 @@ BOOST_AUTO_TEST_CASE(rpc_excessive)
 
     BOOST_CHECK_NO_THROW(CallRPC("getminingmaxblock"));
 
-    BOOST_CHECK_NO_THROW(CallRPC("setminingmaxblock 2000"));
+    // Testing the parsing of input parameters of setexcessive block,
+    // this RPC set the value for EB and AD and expect exaclty 2 unsigned
```
Comment (gandrewstone) [inline src/test/excessiveblock_test.cpp]: spelling of exactly

File: src/test/excessiveblock_test.cpp [hunk 11]
```diff
@@ -184,49 +209,46 @@ BOOST_AUTO_TEST_CASE(check_validator_rule)
...
-    str = ExcessiveBlockValidator(tmpExcessive, NULL, false);
-    BOOST_CHECK(str.empty());
-
-    str = ExcessiveBlockValidator(tmpExcessive, (uint64_t *)42, true);
-    BOOST_CHECK(str.empty());
+    // Tweaks validator is potentially executed twice for every set operation.
+    // The first execution check the validity of the value we want to set the param to.
+    // The second execution happens only if the assignment happened succesfully and
+    // could be used as notification update mechanism.
+    // Tthe first time with the 3rd par (validate) = true and the 2n with validate = false.
```
Comment (gandrewstone) [inline src/test/excessiveblock_test.cpp]: typo: Tthe and 2n -> 2nd

File: src/globals.cpp (hunk 6 above)
Comment (sickpig) [reply src/globals.cpp]: done

File: src/chainparams.cpp (hunk 7 above)
Comment (sickpig) [reply src/chainparams.cpp]: done

File: src/chainparams.cpp (hunk 8 above)
Comment (sickpig) [reply src/chainparams.cpp]: done

File: src/chainparams.cpp (hunk 9 above)
Comment (sickpig) [reply src/chainparams.cpp]: done

File: src/test/excessiveblock_test.cpp (hunk 10 above)
Comment (sickpig) [reply src/test/excessiveblock_test.cpp]: done

File: src/test/excessiveblock_test.cpp (hunk 11 above)
Comment (sickpig) [reply src/test/excessiveblock_test.cpp]: done

**Pull request merged**